src/ci_providers/provider_bitbucket.js

```diff
@@ -0,0 +1,78 @@
+const childProcess = require('child_process')
+const { validateSHA } = require('../helpers/validate.js')
+
+function detect(envs) {
+  return envs.CI && envs.BITBUCKET_BUILD_NUMBER
+}
+
+function _getBuild(inputs) {
+  const { args, envs } = inputs
+  return args.build || envs.BITBUCKET_BUILD_NUMBER
+}
+
+// eslint-disable-next-line no-unused-vars
+function _getBuildURL(inputs) {
+  return ''
+}
+
+function _getBranch(inputs) {
+  const { args, envs } = inputs
+  return args.branch || envs.BITBUCKET_BRANCH || ''
+}
+
+function _getJob(envs) {
+  return envs.BITBUCKET_BUILD_NUMBER
+}
+
+function _getPR(inputs) {
+  const { args, envs } = inputs
+  return args.pr || envs.BITBUCKET_PR_ID || ''
+}
+
+function _getService() {
+  return 'bitbucket'
+}
+
+function getServiceName() {
+  return 'Bitbucket'
+}
+
+function _getSHA(inputs) {
+  const { args, envs } = inputs
+  let commit = envs.BITBUCKET_COMMIT
+
+  if (commit && validateSHA(commit, 12)) {
+    commit = childProcess.execFileSync('git', ['rev-parse', commit])
+  }
+
+  return args.sha || commit
+}
+
+function _getSlug(inputs) {
+  const { args, envs } = inputs
+
+  let slug = ''
+  if (envs.BITBUCKET_REPO_OWNER && envs.BITBUCKET_REPO_SLUG) {
+    slug = `${envs.BITBUCKET_REPO_OWNER}/${envs.BITBUCKET_REPO_SLUG}`
+  }
+  return args.slug || slug
```

Comment:
Be aware that `slug` is sourced from the environment. This is a not bad thing by any means but it's important to make sure this is never provided to something that expects a sanitized string.

Since this project is written in JavaScript rather than TypeScript, tracking down these sources of tainted data may prove to be quite difficult. CodeQL is marketed as doing this kind of tracking but it is probably not mature enough to be completely reliant upon.

It would be worth investigating TypeScript for use in this project—worth its weight in gold, in fact, because an effective TypeScript configuration would catch the most common classes of errors that happen in JavaScript.

Reply:
Indeed, and I'd greatly prefer that to JSDocs. I'd make some _mumble mumble_ about a start date of Nov 21, 2019 and a contributor % https://github.com/codecov/uploader/graphs/contributors, but that's me still making excuses for my less them perfect JS skills lol.

Good catch about the envs though, will validate them.

Reply:
@thomasrockhu what's a good regex mask for a slug? It always has to be `^[\w\d\-\/]+$`, correct?

Reply:
Ugg, no. I need support for international unicode chars as well. 🤔

Reply:
https://github.com/codecov/uploader/issues/187 for TypeScript

Reply:
I'm going to sidebar validating slug to https://github.com/codecov/uploader/issues/188, so I don't scope creep too much. The Bitbucket support is blocking a customer trialing it. Adding a note for myself to see if we are passing an unfiltered slug in the other providers.

Reply:
> Good catch about the envs though, will validate them.

With regards to validating things like slugs (or names, addresses, etc.), unless there is some definite shape to the piece of data (which should be guided by, at most, technical limitations) from which a validation algorithm can be defined, validation may not be necessary.

Often times, it is assumed that there is a need for validation simply on the basis of having a piece of data to validate. For example, what does a valid name look like in the context of an electronic form for a doctor's office? There's not really a definite answer because a person's name can be anything. There's [a great blog post from 2010](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/) that touches upon this kind of area quite well (please excuse the crassness towards the end).

In the context of `slug`, *sanitization* would certainly applicable if the value is ever used in place of a string wherein certain characters must be escaped.

Reply:
Indeed , I'm mixing sanitize and validate, when they are not. As far as validate, there is a technical limit on some of these, and I'd rather it get stopped client side then at the server, though I'm open to discussion on that.

Regarding sanitize, That's similar. For example, how do I make sure that we don't block the customers who use emojis for their branch names? (yes, they do!)

Reply:
> I'd rather it get stopped client side then at the server

It's great that you brought this up.

In a client-server model, the server should be entirely authoritative of information it has ownership of. In terms of implementation, the server would validate everything sent to it by the client that would normally require validation. This is because the server does not have actual control over what the clients do and how information is generated. Server-side validation, however, does have the benefit of reusability across clients. Web, mobile, desktop, etc., can all send the same type of data to the same endpoint and receive the same responses.

The client *can* also perform validation (this is a common pattern) to provide quicker feedback to the user of any invalid fields. To alleviate the code duplication from this strategy, validation code can be modularized in a manner such that both the client and the server use the same functions, etc., although this is often not applicable if the client is written in a different language from to the server. This is actually a major reason why a lot of developers choose to run their code on Node.js.

I believe GitHub, for example, has most of its domain-specific input validation on the server but the web client is able to dynamically request validation of specific fields to show the user as they type (see [github/auto-check-element](https://github.com/github/auto-check-element)). It also uses native HTML input mechanisms to do basic input type validation for certain input fields although those are by no means authoritative.

Reply:
> In a client-server model, the server should be entirely authoritative of information it has ownership of. 

I'm loving this discussion. Am I understanding that, unless it's a direct consumer of a piece of data, a method or system should not care if the data is valid? There's probably a better way to word that, but it makes sense.

On a similar topic, how much validation of it's inputs should a method do? I waver between "throw errors if it isn't good out of the gate" and "trust that the caller knows what I need". Of these two, the first is probably better. Typescript helps a lot here, but it doesn't cover runtime validation.

Reply:
> Am I understanding that, unless it's a direct consumer of a piece of data, a method or system should not care if the data is valid? There's probably a better way to word that, but it makes sense.

Essentially so.

There is tremendous strength in your understanding given the phrase *should not care*. This treads into the territory of architectural design but there is something called separation of concerns that broadly describes this approach.

> On a similar topic, how much validation of it's inputs should a method do? I waver between "throw errors if it isn't good out of the gate" and "trust that the caller knows what I need". Of these two, the first is probably better. Typescript helps a lot here, but it doesn't cover runtime validation.

Let me give an example of a short refactoring:

```js
const { readFile, stat } = require("fs/promises");

// (1)
async function read(path) {
  if (await stat(path)) {
    const contents = await readFile(path);
    
    // (A)
  } else {
    // (B)
  }
  
  // (C)
}

// (2)
async function read(path) {
  const contents = await readFile(path);
  
  // (A)
}
```

Reply:
- Access to the underlying system resources is requested via the `stat` and `readFile` functions
- They are handled by the OS and does not need to occur sequentially as the code would suggest
- Both can throw an `ENOENT: no such file or directory` error